lib/space_saving.c

```diff
@@ -0,0 +1,795 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+/*
+ * This Implementation is similar to the stream summary data structure decribed
+ * in the research paper on space saving algorithm by Ahmed Metwally,
+ * Divyakant Agrawal and Amr El Abbadi.
+ */
+
+/* 
+ * TODO #1 : Add a Tree based data structure such as AVL Tree or
+ * Van Emde Boas Tree to support sublinear bucket update operation.
+ */
+
+/* TODO #2 : Test the written code for correctness */
+
+#include <time.h>
+
+#include "space_saving.h"
+
+/* Set up the parameters of a counter bucket */
+static int
+setup_bucket_params(unsigned int socket_id, int bkt_size, int bkt_id,
+		int ip_ver, struct rte_hash_parameters *bkt_hash_params)
+{
+	int ret;
+
+	int key_len = ip_ver == ETHER_TYPE_IPv4 ?
+			sizeof(struct in_addr) : sizeof(struct in6_addr);
+
+	bkt_hash_params->name = NULL;
+	bkt_hash_params->entries = bkt_size;
+	bkt_hash_params->key_len = key_len;
+	bkt_hash_params->hash_func = DEFAULT_HASH_FUNC;
+	bkt_hash_params->hash_func_init_val = 0;
+	bkt_hash_params->reserved = 0;
+
+	char bkt_name[64];
```

Comment:
Please move this declaration to the beginning of the function, so that all the variables will be put together except for some local variables using only by some blocks of the functions.